symbolic-il2cpp/src/usym.rs

```diff
@@ -0,0 +1,225 @@
+        let size: usize = u16::from_le_bytes([size_bytes[0], size_bytes[1]]).into();
+
+        let start_offset = offset + 2;
+        let end_offset = start_offset + size;
+
+        let string_bytes = self.string_table.get(start_offset..end_offset)?;
+        Some(String::from_utf8_lossy(string_bytes))
+    }
+
+    /// Returns a [`UsymRecord`] at the given index it was stored.
+    ///
+    /// Not that useful, you have no idea what index you want.
+    fn get_record(&self, index: usize) -> Option<UsymSourceRecord> {
+        let raw = self.records.get(index)?;
+        Some(UsymSourceRecord {
+            address: raw.address,
+            symbol: self.get_string(raw.symbol.try_into().unwrap())?,
+            file: self.get_string(raw.file.try_into().unwrap())?,
+            line: raw.line,
+        })
+    }
+
+    /// Lookup the managed code source location for an IL2CPP instruction pointer.
+    fn lookup_source_record(&self, ip: u64) -> Option<UsymSourceRecord> {
+        // TODO: need to subtract the image base to get relative address
+        match self.records.binary_search_by_key(&ip, |r| r.address) {
+            Ok(index) => self.get_record(index),
+            Err(index) => self.get_record(index - 1),
+        }
+    }
+
+    // TODO: Add iterator over records?
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs::File;
+
+    use symbolic_common::ByteView;
+    use symbolic_testutils::fixture;
+
+    use super::*;
+
+    fn empty_usymlite() -> Result<ByteView<'static>> {
+        let file = File::open(fixture("il2cpp/empty.usymlite"))?;
+        ByteView::map_file_ref(&file).map_err(Into::into)
+    }
+
+    #[test]
+    fn test_test() {
+        let file = File::open(
+            "/Users/flub/code/sentry-unity-il2cpp-line-numbers/Builds/iOS/UnityFramework.usym",
+        )
+        .unwrap();
+        let data = ByteView::map_file_ref(&file).unwrap();
+        let usyms = UsymSymbols::parse(&data).unwrap();
+
+        println!("id: {}", usyms.get_string(0x02).unwrap());
+        println!("name: {}", usyms.get_string(0x24).unwrap());
+        println!("os: {}", usyms.get_string(0x34).unwrap());
+        println!("arch: {}", usyms.get_string(0x39).unwrap());
+
+        println!("{:?}", usyms.get_record(0));
+
+        let record = usyms.lookup_source_record(2883352);
+        println!("record: {record:?}");
+
+        panic!("boom");
+    }
+
+    #[test]
+    fn test_sorted_addresses() {
+        let file = File::open(
+            "/Users/flub/code/sentry-unity-il2cpp-line-numbers/Builds/iOS/UnityFramework.usym",
+        )
+        .unwrap();
+        let data = ByteView::map_file_ref(&file).unwrap();
+        let usyms = UsymSymbols::parse(&data).unwrap();
+
+        let mut last_address = usyms.records[0].address;
+        for i in 1usize..usyms.header.record_count as usize {
+            assert!(usyms.records[i].address > last_address);
+            last_address = usyms.records[0].address;
```

Comment:
Should this be set to `usyms.records[i].address`, or is the test meant to be constantly checking against the first record's address?